src/network/service.rs

```diff
@@ -247,12 +261,41 @@ where
         });
         let response = self
             .libp2p
-            .request(now, target, 1 + chain_index * 3, request_data)
+            .request(
+                now,
+                target,
+                self.protocol_index(chain_index, 0),
+                request_data,
+            )
             .map_err(BlocksRequestError::Request)
             .await?;
         protocol::decode_block_response(&response).map_err(BlocksRequestError::Decode)
     }
 
+    pub async fn grandpa_warp_sync_request(
+        &self,
+        now: TNow,
+        target: peer_id::PeerId,
+        chain_index: usize,
+        begin_hash: [u8; 32],
+    ) -> Result<Vec<GrandpaWarpSyncResponseFragment>, GrandpaWarpSyncRequestError> {
+        use parity_scale_codec::{Compact, Decode, Encode};
+        let request_data = GrandpaWarpSyncRequest { begin: begin_hash }.encode();
```

Comment:
The request is literally just the hash, so `being_hash.to_vec()` is IMO clearer.